### DISCUSSION TITLE
TX23 anemometer does not work in ESP32 modules

### DISCUSSION THREAD

**Discussion by valdus2** [arendst/Tasmota#21390] [Support request]
Tried esp32-devkit ESP32-D0WD-V3 v3.0 and ESP32-D0WD-V3 v3.1, also ESP32D.

Compiled either release or development from 13.4.0.4 to 13.4.1.2.

Configuring different GPIOs (13,18,23,26,34).

Connected it with and without external 10k pull-up resistor.

All I get is the below screen for 3 seconds and then it disappears.
![1](https://github.com/arendst/Tasmota/assets/150083707/7e4836e8-a57f-4120-bdb3-a5c66b4bfcfd)

In GPIO viewer I can see the defined gpio alternating between high and low every about 4 seconds.

This is a weblog4 from latest release compiled with tx23:
```
00:00:00.003 HDW: ESP32-D0WD-V3 v3.0 
00:00:00.007 UFS: FlashFS mounted with 304 kB free
00:00:00.013 CFG: Loaded from File, Count 67
00:00:00.023 QPC: Count 1
00:00:00.025 CFG: CR 414/699, Busy 0
00:00:00.030 TYA: Active=0
00:00:00.036 ROT: Mode 1
00:00:00.047 BRY: GC from 3692 to 3492 bytes, objects freed 4/20 (in 0 ms) - slots from 36/61 to 28/61
00:00:00.068 CFG: No '*.autoconf' file found
00:00:00.074 BRY: GC from 5852 to 4446 bytes, objects freed 6/40 (in 0 ms) - slots from 64/91 to 41/91
00:00:00.074 BRY: Berry initialized, RAM used 4446 bytes
00:00:00.082 BRY: No 'preinit.be'
00:00:00.085 SRC: Restart
00:00:00.089 Project tasmota - Tasmota Version 13.4.0(DigiESP32meteo_v24_1b)-2_0_14(2024-05-08T21:31:44)
00:00:00.090 ETH: No ETH MDC and ETH MDIO GPIO defined
00:00:00.105 BRY: No 'autoexec.be'
00:00:00.246 WIF: Checking connection...
00:00:00.247 WIF: Attempting connection...
00:00:00.882 WIF: Connecting to AP1 default3 Channel 10 BSSId 14:EB:B6:B7:A4:FE in mode 11n as tasmota-75A3F8-1016...
00:00:02.213 WIF: IPv4 192.168.0.64, mask 255.255.255.0, gateway 192.168.0.254
00:00:03.424 WIF: IPv6 Local fe80::e25a:1bff:fe75:a3f8%st1
00:00:03.447 WIF: Checking connection...
00:00:03.447 WIF: Connected
00:00:03.706 HTP: Web server active on tasmota-75A3F8-1016 with IP address 192.168.0.64
00:00:03.999 NTP: Sync time...
00:00:04.059 WIF: DNS resolved '2.pool.ntp.org' (162.159.200.1) in 58 ms
00:00:04.085 RTC: UTC 2024-05-08T21:49:09Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
22:49:09.000 RTC: Synced by NTP
22:49:09.613 MQT: Attempting connection...
22:49:09.616 WIF: DNS resolved '192.168.0.5' (192.168.0.5) in 1 ms
22:49:09.645 MQT: Connected
22:49:09.650 MQT: tele/tasmota_75A3F8/LWT = Online (retained)
22:49:09.652 MQT: cmnd/tasmota_75A3F8/POWER = 
22:49:09.653 MQT: Subscribe to cmnd/tasmota_75A3F8/#
22:49:09.655 MQT: Subscribe to cmnd/tasmotas/#
22:49:09.656 MQT: Subscribe to cmnd/DVES_75A3F8_fb/#
22:49:09.660 MQT: tele/tasmota_75A3F8/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.4.0(DigiESP32meteo_v24_1b)","FallbackTopic":"cmnd/DVES_75A3F8_fb/","GroupTopic":"cmnd/tasmotas/"}}
22:49:09.678 MQT: tele/tasmota_75A3F8/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-75A3F8-1016","IPAddress":"192.168.0.64","IP6Global":"","IP6Local":"fe80::e25a:1bff:fe75:a3f8%st1"}}
22:49:09.693 MQT: tele/tasmota_75A3F8/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":39}}
22:49:09.708 TFS: File 'autoexec.bat' not found
22:49:11.964 QPC: Reset
22:49:12.605 HTP: Main Menu
22:49:13.985 APP: Boot Count 39
22:49:13.988 BRY: GC from 6374 to 4212 bytes, objects freed 26/45 (in 1 ms) - slots from 91/122 to 50/122
22:49:13.998 MQT: tele/tasmota_75A3F8/STATE = {"Time":"2024-05-08T22:49:13","Uptime":"0T00:00:09","UptimeSec":9,"Heap":151,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"default3","BSSId":"14:EB:B6:B7:A4:FE","Channel":10,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:04"}}
22:49:14.430 CFG: Saved, Count 68, Bytes 4096
22:49:19.020 MQT: tasmota/discovery/E05A1B75A3F8/config = {"ip":"192.168.0.64","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-75A3F8-1016","mac":"E05A1B75A3F8","md":"ESP32-DevKit","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"13.4.0","t":"tasmota_75A3F8","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1} (retained)
22:49:19.027 MQT: tasmota/discovery/E05A1B75A3F8/sensors = {"sn":{"Time":"2024-05-08T22:49:19"},"ver":1} (retained)
22:49:28.450 WIF: Checking connection...
22:49:33.958 BRY: GC from 4567 to 4212 bytes, objects freed 1/45 (in 0 ms) - slots from 51/122 to 50/122
22:49:33.969 MQT: tele/tasmota_75A3F8/STATE = {"Time":"2024-05-08T22:49:33","Uptime":"0T00:00:29","UptimeSec":29,"Heap":150,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"default3","BSSId":"14:EB:B6:B7:A4:FE","Channel":10,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:04"}}
22:49:34.314 HTP: Tools
22:49:35.268 HTP: Console
22:49:48.475 WIF: Checking connection...
22:49:53.982 BRY: GC from 4573 to 4212 bytes, objects freed 1/45 (in 0 ms) - slots from 51/122 to 50/122
22:49:53.992 MQT: tele/tasmota_75A3F8/STATE = {"Time":"2024-05-08T22:49:53","Uptime":"0T00:00:49","UptimeSec":49,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"default3","BSSId":"14:EB:B6:B7:A4:FE","Channel":10,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:04"}}
22:50:08.494 WIF: Checking connection...
22:50:14.002 BRY: GC from 4581 to 4212 bytes, objects freed 1/45 (in 0 ms) - slots from 50/122 to 50/122
22:50:14.011 MQT: tele/tasmota_75A3F8/STATE = {"Time":"2024-05-08T22:50:13","Uptime":"0T00:01:09","UptimeSec":69,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"default3","BSSId":"14:EB:B6:B7:A4:FE","Channel":10,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:04"}}
22:50:28.462 WIF: Checking connection...
22:50:33.970 BRY: GC from 4548 to 4212 bytes, objects freed 1/45 (in 0 ms) - slots from 50/122 to 50/122
22:50:33.979 MQT: tele/tasmota_75A3F8/STATE = {"Time":"2024-05-08T22:50:33","Uptime":"0T00:01:29","UptimeSec":89,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":45},"Wifi":{"AP":1,"SSId":"default3","BSSId":"14:EB:B6:B7:A4:FE","Channel":10,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:04"}}
```

Everything is working ok, **ONLY** with nodeMCU(_ESP8266_) and wemos d1 mini (_ESP8266_).

Any help would be appreciated becouse I'm really out of ideas.
Thank you in advance.

**Comment by valdus2** [comment]
Should I open an issue? I don't know what to do.

**Reply by arendst** [reply]
Leave it here for someone with a device and driver knowledge can start testing and debugging.

I guess the device has never been tested on ESP32....

**Reply by valdus2** [reply]
So, just waiting and hoping......Thank you **arendst**.

**Reply by harryth1** [reply]
Hello,
I have exactly the same issue as valdus2.
I also have tried on different GPIOs and ESP32 based Board as well as on NodeMCU ESP8266-12F based board.
After Reboot the different wind parameters are displayed for ~3 sec and then they disappear.
I checked with oscilloscope and can see the answer from TX23 after the GPIO was pulled low by the sw, which occures every 4 sec.
Is there a chance that someone, who has experience with this driver can have a look on it.
If wanted, I can support with testing.

Thanks a lot in advance

**Comment by harryth1** [comment]
Hello,
is there a chance, that the above described behaviour will be investigated by an experienced driver developer?
As already offered, I could support with testing.

Thank you in advance

**Reply by curzon01** [reply]
My anemometer and ESP are outside, and we have -7 °C, which is no fun at the moment.

You can do the following in advance:

- Does the signal GPIO have an additional pull-up and what is the value?
- Is the wind meter supplied with 3.3V or 5V? In the latter case, is there a voltage divider/level shifter?
- Which GPIO is used?

First, use the ESP8266 (there could be a problem with the ESP32) and the current Tasmota 15.x.x, compile it with #define DEBUG_TASMOTA_SENSOR, otherwise Weblog 4 will not display more than Weblog 2.

**Comment by harryth1** [comment]
Hello,
thanks for taking care ...

I am using ESP-07 Module 
The TX23 is connected to GPIO16 and a 10k pullup is used
The TX23 is supplied by 3.3V from the USB to Serial converter
I have still the long cable since I will need it once it will be installed on the roof of my house

Here is the Log from restart with weblog 4:

00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F7, Count 23
00:00:00.055 QPC: Count 1
00:00:00.057 CFG: CR 353/699, Busy 0
00:00:00.061 TYA: Active=0
00:00:00.064 ROT: Mode 1
00:00:00.065 SRC: Restart
00:00:00.066 Project tasmota - Tasmota Version 15.2.0.2(tasmota)-2.7.8(2026-01-08T15:46:36)
00:00:00.067 TX23: reset statistics
00:00:00.074 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
00:00:00.075 TX23: dir stat - counter=1, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
00:00:00.219 WIF: Checking connection...
00:00:00.219 WIF: Attempting connection...
00:00:01.001 WIF: Connecting to AP1 wlan@wg-th.de Channel 11 BSSId 7C:FF:4D:14:B4:7E in mode 11n as tasmota-CFEAF3-2803...
00:00:01.004 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
00:00:01.005 TX23: dir stat - counter=2, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
00:00:01.765 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
00:00:01.766 TX23: dir stat - counter=3, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
00:00:02.215 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
00:00:02.216 TX23: dir stat - counter=4, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
00:00:02.514 WIF: Checking connection...
00:00:02.515 WIF: Attempting connection...
00:00:03.215 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
00:00:03.217 TX23: dir stat - counter=5, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
00:00:03.514 WIF: Checking connection...
00:00:03.515 WIF: Connected
00:00:03.767 HTP: Web server active on tasmota-CFEAF3-2803 with IP address 192.168.0.33
00:00:04.215 WIF: Sending Gratuitous ARP
00:00:04.217 NTP: Sync time...
00:00:04.252 WIF: DNS resolved '2.pool.ntp.org' (78.46.56.170) in 34 ms
00:00:04.285 RTC: UTC 2026-01-08T14:56:24Z, DST 2026-03-29T02:00:00, STD 2026-10-25T03:00:00
15:56:24.001 RTC: Synced by NTP
15:56:24.003 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
15:56:24.004 TX23: dir stat - counter=6, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
15:56:24.511 RSL: INFO1 = {"Info1":{"Module":"ESP-07","Version":"15.2.0.2(tasmota)","FallbackTopic":"cmnd/DVES_CFEAF3_fb/","GroupTopic":"cmnd/tasmotas/"}}
15:56:24.515 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-CFEAF3-2803","IPAddress":"192.168.0.33"}}
15:56:24.526 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":11}}
15:56:25.229 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
15:56:25.231 TX23: dir stat - counter=7, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
15:56:26.371 QPC: Reset
15:56:26.373 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
15:56:26.375 TX23: dir stat - counter=8, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
15:56:27.231 TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0
15:56:27.233 TX23: dir stat - counter=9, actint=0, avgint=0, avg=0.0 (cosx=1.0, siny=0.0), min 0, max 0
15:56:28.245 APP: Boot Count 11
15:56:28.246 TX23: not available
15:56:28.434 CFG: Saved to flash at F6, Count 24, Bytes 4096
15:56:29.246 RSL: STATE = {"Time":"2026-01-08T15:56:29","Uptime":"0T00:00:11","UptimeSec":11,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"wlan@wg-th.de","BSSId":"7C:FF:4D:14:B4:7E","Channel":11,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":1,"Downtime":"0T00:00:05"},"Hostname":"tasmota-CFEAF3-2803","IPAddress":"192.168.0.33"}
15:56:29.270 TX23: reset statistics
15:56:29.271 TX23: not available
15:56:30.234 TX23: not available
15:56:31.235 TX23: not available
15:56:32.237 TX23: not available
15:56:33.213 TX23: not available
15:56:34.234 RSL: config = {"ip":"192.168.0.33","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-CFEAF3-2803","mac":"083A8DCFEAF3","md":"ESP-07","ty":0,"if":0,"cam":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"15.2.0.2","t":"tasmota_CFEAF3","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1}
15:56:34.240 RSL: sensors = {"sn":{"Time":"2026-01-08T15:56:34"},"ver":1}
15:56:34.241 TX23: not available
15:56:35.246 TX23: not available
15:56:36.210 TX23: not available


Below a screenshot shortly after reboot:
(After ~3 seconds the wind menue disappeares)
<img width="1512" height="1512" alt="grafik" src="https://github.com/user-attachments/assets/546658c4-b3af-4cd9-b8be-9251758ad763" />

**Reply by curzon01** [reply]
GPIO16 is not a good idea, see also [TX23 Tasmota Setting](https://tasmota.github.io/docs/TX2x/#tasmota-settings) in docs

**Reply by sfromis** [reply]
Beware that some USB/serial adapters really can't even supply good power for ESP8266 with Wifi. Not saying "that's it", just that having good power all the way is a really important thing to be extra sure of. *At least* 150 mA of 3.3V power. Still, the point about gpio 16 as a plausible troublemaker is a good one.

**Reply by curzon01** [reply]
All the lines `TX23: sa=0x00 sb=0 (0x00), sc=0 (0x000), sd=0x00, se=0, sf=0`  tells us that after the TX23 start signal is initiated, no serial information comes through the interface – not even during the first three seconds. 

The reason why the display on the GUI only disappears after 3 seconds is because the driver decides to switch off the display after 3 seconds without data – not because 3 seconds of data have come in and then no more. The debug log clearly states this.

In addition to the incorrect GPIO, sufficient power supply is also a crucial factor.

**Comment by harryth1** [comment]
I have changed to GPIO12 with an external 10k pullup, powered by the USB to Serial converter (FTDI232RL based) and now it works as expected; the values of windspeed and wind direction change according to the sensor's movement.
Nevertheless I have noticed, that the USBtoSerial converter can only supply 50 mA at 3V3OUT pin. A lab power supply showed a supply current of 68 to 80 mA to the module, so for stable operation the converter is definitely not suited.

So thank you very much for the hints, it was my fault not to read the comments carefully :-(

The original idea nevertheless was to run the TX23 anemomenter on an ESP32C3-S5 based module with Ethernet and POE power supply.
(ETH01-Evo with ETH01-POE). The TX23 would be powered by an onboard DC-DC converter which can deliver 2.5A at 3.3V. So the power supply should be ok. The TX23 is connected to GPIO4 in this scenario and has also a 10k pullup to 3.3V.
Here also I have the issue, that only zero values are shown and the display disappaears after 3 sec.
What info do you need from me for further debugging?

Once again thanks for your time you spend on this.

**Reply by curzon01** [reply]
> What info do you need from me for further debugging?

I first need to test this locally with my TX23 and an ESP32, but unfortunately that will have to wait as the weather is not permitting it at the moment, sorry
